runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -1235,7 +1235,15 @@ void loadFieldWatchSnippet(TR::CodeGenerator *cg, TR::Node *node, TR::Snippet *d
       {
       if (beginIndex<LOWER_IMMED || beginIndex>UPPER_IMMED)
          {
-         generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addis, node, scratchReg, cg->getTOCBaseRegister(), HI_VALUE(beginIndex));
+         if (0x00008000 == HI_VALUE(beginIndex))
+            {
```

Comment:
This is wrong. It deals with legitimate offsets such as 0x80000000 wrongly, turning them into positive offsets. You need to test specifically the other situations.

Reply:
If `beginIndex` is equal to `0x80000000` then `HI_VALUE(beginIndex)` will become `0xFFFF8000` so it won't trigger the if statement.

`HI_VALUE(beginIndex)` will only become `0x00008000` when the upper 16 bits of `beginIndex` are equal to `0x7FFF` and the highest bit of the lower 16 bits is set.

So `beginIndex` has to look like this in binary:
`0111 1111 1111 1111 1XXX XXXX XXXX XXXX`

`HI_VALUE` will first shift right by 16 bits:
`0000 0000 0000 0000 0111 1111 1111 1111`
Then add 1 because the highest bit of the original lowest 16 bits was set:
`0000 0000 0000 0000 1000 0000 0000 0000` = `0x00008000`

Reply:
fair enough.